discovery/validation_barrier.go

```diff
@@ -27,3 +58,446 @@ type ValidationBarrier interface {
+// based on any pending uncompleted dependent jobs. If this job doesn't have an
+// active dependent, then this function will return immediately.
+func (g *GossipBarrier) WaitForDependants(nMsg *networkMsg) error {
+	// Get the signals for this job.
+	allow, deny, jobDesc := g.getSignals(nMsg)
+
+	// If the signal is nil, it means either the job is not a dependent
+	// type, or it doesn't have a dependency signal. Either way, we can
+	// return early.
+	if allow == nil {
+		return nil
+	}
+
+	log.Debugf("Waiting for dependent on %s", jobDesc)
+
+	// If we do have an active job, then we'll wait until either the signal
+	// is closed, or the set of jobs exits.
+	select {
+	case <-g.quit:
+		return errVBarrierShuttingDown
+
+	case <-deny:
+		log.Debugf("Signal deny for %s", jobDesc)
+		return errParentValidationFailed
+
+	case <-allow:
+		log.Tracef("Signal allow for %s", jobDesc)
+		return nil
+	}
+}
+
+// SignalDependants will allow/deny any jobs that are dependent on this job so
+// they can continue execution. The semaphores are released to empty job slots
+// for future jobs. If the job doesn't have any dependants, then this function
+// will exit immediately.
+//
+// NOTE: This MUST be called after InitJobDependencies is called.
+func (g *GossipBarrier) SignalDependants(nMsg *networkMsg, allow bool) {
+	// Get the sender node's public key.
+	var relayingNode route.Vertex
+	copy(relayingNode[:], nMsg.source.SerializeCompressed())
+
+	// Release the semaphore when job is finished.
+	defer g.releaseSemaphore(relayingNode)
+
+	g.Lock()
+	defer g.Unlock()
+
+	log.Tracef("Signaling dependants for %T, allow=%v", nMsg.msg, allow)
+
+	switch msg := nMsg.msg.(type) {
+	// If we've just finished executing a ChannelAnnouncement, then we'll
+	// close out the signal, and remove the signal from the map of active
+	// ones. This will allow/deny any dependent jobs to continue execution.
+	case *lnwire.ChannelAnnouncement:
+		finSignals, ok := g.chanAnnFinSignal[msg.ShortChannelID]
+
+		// Exit early if there's no dependant.
+		if !ok {
+			return
+		}
+
+		// If this is allow signal, we'll send it to all dependants and
+		// remove it from the map.
+		if allow {
+			close(finSignals.allow)
+			delete(g.chanAnnFinSignal, msg.ShortChannelID)
+
+			return
+		}
+
+		// If this is a deny signal, check if we have any dependants.
+		// If not we'll exit early.
+		denySignal, ok := finSignals.deny[relayingNode]
+		if !ok {
+			return
+		}
+
+		// Otherwise send the deny signal to the specific node and
+		// remove it from the map.
+		//
+		// NOTE: we cannot remove it from `chanAnnFinSignal` yet as
+		// it's shared by other relaying nodes. We will need it here if
+		// a different node is sending signal deny.
+		close(denySignal)
+
+		// TODO(yy): instead of removing the deny signal, we could
+		// count how many deny signals have been sent from this
+		// relaying node and further ban this misbehaving peer.
```

Comment:
> And banning by node pubkey is not very effective since node pubkeys have no cost.

Indeed. If we optionally require gossiping only if we have an open channel then it will be costly to carry the attack.

> it's possible that they have some implementation bug that we end up banning.

If we know they are buggy then we should disconnect?

I think a deny signal means the node did something wrong and we need to do something about it. Or at least, let the user know this peer is misbehaving if many denies are sent.